File: kotlin/conformance/src/test/kotlin/com/basecamp/sdk/conformance/DelayGapsTest.kt
package com.basecamp.sdk.conformance

import kotlin.test.Test
import kotlin.test.assertContains
import kotlin.test.assertEquals
import kotlin.test.assertNull

/**
 * Bounds and every-gap contract for the `delayBetweenRequests` assertion.
 *
 * Each case names a behavior that regressed on #563/#568: an assertion that
 * looked like it covered a timing gap and did not.
 */
class DelayGapsTest {
    /** Builds request timestamps from successive gaps in milliseconds. */
    private fun times(vararg gapsMs: Long): List<Long> {
        val out = mutableListOf(0L)
        for (ms in gapsMs) out.add(out.last() + ms)
        return out
    }

    @Test
    fun `omitted index catches a later failing gap`() {
        // Three runners measured gap 0 and stopped, so a second backoff that
        // never happened passed unnoticed.
        assertContains(checkDelayGaps(times(1000, 5), 500, null)!!, "at gap 1")
    }

    @Test
    fun `omitted index passes when every gap clears the minimum`() {
        assertNull(checkDelayGaps(times(1000, 2000, 800), 500, null))
    }

    @Test
    fun `omitted index fails when there are no gaps at all`() {
        // An "every gap" rule with no gaps left must not wave the run through:
        // a fully dropped retry lands exactly here.
        assertEquals(
            "Expected a delay between requests, but only 1 request(s) were made",
            checkDelayGaps(times(), 500, null),
        )
    }

    @Test
    fun `named gap fails when the run never produced it`() {
        assertEquals(
            "Expected a delay at gap 1, but only 2 request(s) were made",
            checkDelayGaps(times(1000), 500, 1),
        )
    }

    @Test
    fun `named gap fails on a single-request run`() {
        assertEquals(
            "Expected a delay at gap 0, but only 1 request(s) were made",
            checkDelayGaps(times(), 500, 0),
        )
    }

    @Test
    fun `negative gap index is rejected`() {
        // Rejected categorically, not wrapped to the end the way
        // headerPresent's index is.
        assertEquals(
            "delayBetweenRequests gap index must be non-negative, got -1",
            checkDelayGaps(times(1000, 2000), 500, -1),
        )
    }

    @Test
    fun `Int MAX_VALUE gap index fails without overflowing`() {
        // `gap + 1 >= size` computes the addition first, so Int.MAX_VALUE wraps
        // negative and sails through the guard into an out-of-range read.
        assertContains(
            checkDelayGaps(times(1000, 2000), 500, Int.MAX_VALUE)!!,
            "Expected a delay at gap",
        )
    }

    @Test
    fun `zero minimum still asserts that the gap exists`() {
        // A zero minimum is trivially met; the EXISTENCE requirement is not.
        assertEquals(
            "Expected a delay between requests, but only 1 request(s) were made",
            checkDelayGaps(times(), 0, null),
        )
        assertEquals(
            "Expected a delay at gap 0, but only 1 request(s) were made",
            checkDelayGaps(times(), 0, 0),
        )
        assertNull(checkDelayGaps(times(5), 0, null))
    }

    @Test
    fun `named gap passes when it clears the minimum`() {
        assertNull(checkDelayGaps(times(5, 2000), 500, 1))
    }

    @Test
    fun `named gap fails when it is below the minimum`() {
        assertContains(checkDelayGaps(times(2000, 5), 500, 1)!!, "at gap 1")
    }
}

File: conformance/runner/ruby/delay_gaps_test.rb
# frozen_string_literal: true

# Bounds and every-gap contract for the delayBetweenRequests assertion.
#
# Each case names a behavior that regressed on #563/#568: an assertion that
# looked like it covered a timing gap and did not.
#
# Run: `bundle exec ruby delay_gaps_test.rb`

require "minitest/autorun"
require_relative "runner"

class DelayGapsTest < Minitest::Test
  def test_omitted_index_catches_a_later_failing_gap
    # Three runners measured gap 0 and stopped, so a second backoff that never
    # happened passed unnoticed.
    failure = DelayGaps.check([ 1000, 5 ], 500, nil)

    assert_includes failure, "at gap 1"
  end

  def test_omitted_index_passes_when_every_gap_clears_the_minimum
    assert_nil DelayGaps.check([ 1000, 2000, 800 ], 500, nil)
  end

  def test_omitted_index_fails_when_there_are_no_gaps_at_all
    # An "every gap" rule with no gaps left must not wave the run through: a
    # fully dropped retry lands exactly here.
    assert_equal "Expected a delay between requests, but only 1 request(s) were made",
      DelayGaps.check([], 500, nil)
  end

  def test_named_gap_fails_when_the_run_never_produced_it
    assert_equal "Expected a delay at gap 1, but only 2 request(s) were made",
      DelayGaps.check([ 1000 ], 500, 1)
  end

  def test_named_gap_fails_on_a_single_request_run
    assert_equal "Expected a delay at gap 0, but only 1 request(s) were made",
      DelayGaps.check([], 500, 0)
  end

  def test_negative_gap_index_is_rejected
    # Rejected categorically, not wrapped to the end the way headerPresent's
    # index is: there is no sensible "last gap" when the point is to name one.
    assert_equal "delayBetweenRequests gap index must be non-negative, got -1",
      DelayGaps.check([ 1000, 2000 ], 500, -1)
  end

  def test_enormous_gap_index_fails_rather_than_raising
    assert_includes DelayGaps.check([ 1000, 2000 ], 500, 2**63 - 1), "Expected a delay at gap"
  end

  def test_named_gap_passes_when_it_clears_the_minimum
    assert_nil DelayGaps.check([ 5, 2000 ], 500, 1)
  end

  def test_named_gap_fails_when_it_is_below_the_minimum
    assert_includes DelayGaps.check([ 2000, 5 ], 500, 1), "at gap 1"
  end

  def test_zero_minimum_still_asserts_that_the_gap_exists
    # A `min` of zero (or an absent one, which the caller defaults to zero)
    # still requires the gap to EXIST. Gating the call on the value's presence
    # would degrade the assertion to nothing.
    [ 0, nil ].each do |missing|
      assert_equal "Expected a delay between requests, but only 1 request(s) were made",
        DelayGaps.check([], missing, nil)
      assert_equal "Expected a delay at gap 0, but only 1 request(s) were made",
        DelayGaps.check([], missing, 0)
      assert_nil DelayGaps.check([ 5 ], missing, nil)
    end
  end
end
